File: 0221-Maximal Square/c++/solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        if(matrix.empty()) return 0;
        const int x = matrix.size();
        const int y = matrix[0].size();
        
        int ans = 0;
        for(int i = 0; i < x; ++i) {
            for(int j = 0; j < y; ++j) {
                matrix[i][j] -= '0';
                
                if(i!=0 && j!=0 && matrix[i][j]==1) {
                    matrix[i][j] = min(min(matrix[i-1][j], matrix[i][j-1]), matrix[i-1][j-1]) + matrix[i][j];
                }
                if(matrix[i][j] > ans) ans = matrix[i][j];
            }
        }
        return ans*ans;
    }
};

// Runtime: 32 ms, faster than 99.48% of C++ online submissions for Maximal Square.
// Memory Usage: 11.2 MB, less than 80.82% of C++ online submissions for Maximal Square.

File: 0221-Maximal Square/golang/solution.go
package golang

func min(x, y byte) byte {
	if x > y {
		return y
	}
	return x
}

func maximalSquare(matrix [][]byte) int {
	if len(matrix) == 0 {
		return 0
	}

	var sum int = 0
	for i := 0; i < len(matrix); i++ {
		for j := 0; j < len(matrix[0]); j++ {
			matrix[i][j] -= '0'
			if i != 0 && j != 0 && matrix[i][j] == 1 {
				matrix[i][j] = min(min(matrix[i-1][j], matrix[i][j-1]), matrix[i-1][j-1]) + matrix[i][j]
			}

			if sum < int(matrix[i][j]) {
				sum = int(matrix[i][j])
			}
		}
	}
	return sum * sum
}

// Runtime: 0 ms, faster than 100.00% of Go online submissions for Maximal Square.
// Memory Usage: 3.4 MB, less than 94.48% of Go online submissions for Maximal Square.
